sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/ReleaseDownloadListener.java

```diff
@@ -0,0 +1,196 @@
+import static com.microsoft.appcenter.distribute.DistributeConstants.PREFERENCE_KEY_DOWNLOADED_DISTRIBUTION_GROUP_ID;
+import static com.microsoft.appcenter.distribute.DistributeConstants.PREFERENCE_KEY_DOWNLOADED_RELEASE_HASH;
+import static com.microsoft.appcenter.distribute.DistributeConstants.PREFERENCE_KEY_DOWNLOADED_RELEASE_ID;
+import static com.microsoft.appcenter.distribute.DistributeConstants.PREFERENCE_KEY_DOWNLOAD_STATE;
+import static com.microsoft.appcenter.distribute.DistributeConstants.PREFERENCE_KEY_DOWNLOAD_TIME;
+import static com.microsoft.appcenter.distribute.InstallerUtils.getInstallIntent;
+
+/**
+ * Listener for downloading progress.
+ */
+class ReleaseDownloadListener implements ReleaseDownloader.Listener {
+
+    /**
+     * Context.
+     */
+    private final Context mContext;
+
+    /**
+     * Private field to store information about release we are currently working with.
+     */
+    private final ReleaseDetails mReleaseDetails;
+
+    ReleaseDownloadListener(@NonNull Context context, @NonNull ReleaseDetails releaseDetails) {
+        mContext = context;
+        mReleaseDetails = releaseDetails;
+    }
+
+    /**
+     * Last download progress dialog that was shown.
+     * Android 8 deprecates this dialog but only reason is that they want us to use a non modal
+     * progress indicator while we actually use it to be a modal dialog for forced update.
+     * They will always keep this dialog to remain compatible but just mark it deprecated.
+     */
+    @SuppressWarnings({"deprecation", "RedundantSuppression"})
+    private android.app.ProgressDialog mProgressDialog;
+
+    @Override
+    public void onStart(long enqueueTime) {
+        AppCenterLog.debug(LOG_TAG, String.format(Locale.ENGLISH, "Start download %s (%d) update.",
+                mReleaseDetails.getShortVersion(), mReleaseDetails.getVersion()));
+        SharedPreferencesManager.putInt(PREFERENCE_KEY_DOWNLOAD_STATE, DOWNLOAD_STATE_ENQUEUED);
+        SharedPreferencesManager.putLong(PREFERENCE_KEY_DOWNLOAD_TIME, enqueueTime);
+    }
+
+    @Override
+    @WorkerThread
+    public boolean onProgress(final long currentSize, final long totalSize) {
+        AppCenterLog.verbose(LOG_TAG, String.format(Locale.ENGLISH, "Downloading %s (%d) update: %d KiB / %d KiB",
+                mReleaseDetails.getShortVersion(), mReleaseDetails.getVersion(),
+                currentSize / KIBIBYTE_IN_BYTES, totalSize / KIBIBYTE_IN_BYTES));
+
+        /* If file size is known update downloadProgress bar. */
+        if (mProgressDialog != null && totalSize >= 0) {
+            HandlerUtils.runOnUiThread(new Runnable() {
+
+                @Override
+                @SuppressWarnings({"deprecation", "RedundantSuppression"})
+                public void run() {
+
+                    /* When we switch from indeterminate to determinate */
+                    if (mProgressDialog.isIndeterminate()) {
```

Comment:
state can change after runOnUiThread and thus the progress dialog might be null now. This was managed before on develop.